File: php/src/Commands.php
<?php

namespace App;

include_once './Farm.php';

/**
 * Class that works as a kind of UI
 *
 * @author brayvasq
 */
class Commands {

    private $farm;
    private $data;
    private $dataLength;

    /**
     * Class constructor
     *
     * @param data The input array
     */
    public function __construct($data) {
        $this->farm = new Farm();
        $this->data = $data;
        $this->dataLength = count($this->data);
    }

    /**
     * Obtain the command and process the input data
     */
    public function process() {
        $command = $this->getCommand();
        if ($command != null) {
            $this->verifyCommand($command);
        } else {
            $this->help();
        }
    }

    /**
     * Returns the command
     *
     * @return The first item in the input array, if it's given
     */
    private function getCommand() {
        return $this->dataLength > 1 ? $this->data[1] : null;
    }

    /**
     * Identify the specific action to execute and call the respective action
     *
     * @param command Command obtained from getCommand() method
     */
    private function verifyCommand(String $command) {
        if (in_array($command, array("list", "l"))) {
            $this->listCommand();
        } else if (in_array($command, array("create", "c"))) {
            if ($this->createCommand()) {
                $this->listCommand();
            }
        } else if (in_array($command, array("delete", "d"))) {
            if ($this->deleteCommand()) {
                $this->listCommand();
            }
        } else if (in_array($command, array("search", "s"))) {
            $this->searchCommand();
        } else if (in_array($command, array("food", "f"))) {
            $this->foodCommand();
        } else if (in_array($command, array("wash", "w"))) {
            $this->washCommand();
        } else if (in_array($command, array("alive", "a"))) {
            $this->aliveCommand();
        } else if (in_array($command, array("help", "h"))) {
            $this->help();
        } else {
            $this->help();
        }
    }

    /**
     * Process the [list | l] command
     */
    private function listCommand() {
        $this->farm->listAnimals();
    }

    /**
     * Process the [create | c] command
     *
     * @return true, if the animal was created
     */
    private function createCommand() {
        $name = $this->getParam("name=", "([a-zA-Z]+)");
        $type = $this->getParam("type=", "([a-zA-Z]+)");

        $created = false;

        if ($name != null && !empty($name)) {
            $created = $this->farm->createAnimal($name, $type);
        } else {
            echo "Error!: A name is required to create an animal" . PHP_EOL;
            $this->createCommandHelp();
        }

        return $created;
    }

    /**
     * Process the [delete | c] command
     *
     * @return true if the animal was deleted
     */
    private function deleteCommand() {
        $name = $this->getParam("name=", "([a-zA-Z]+)");
        
        $deleted = false;
        
        if ($name != null && !empty($name)) {
            $deleted = $this->farm->deleteAnimal($name);
        } else {
            echo "Error!: A name is required to delete a animal" . PHP_EOL;
            $this->deleteCommandHelp();
        }
        
        return $deleted;
    }

    /**
     * Process the [search | s] command
     */
    private function searchCommand() {
        $name = $this->getParam("name=", "([a-zA-Z]+)");

        if ($name != null) {
            $this->farm->search($name);
        } else {
            echo "Error!: A name is required to search a animal" . PHP_EOL;
            $this->searchCommandHelp();
        }
    }
    
    /**
     * Process the [food | f] command
     */
    private function foodCommand(){
        $name = $this->getParam("name=", "([a-zA-Z]+)");

        if ($name != null) {
            $this->farm->food($name);
        } else {
            echo "Error!: A name is required to give food to a animal". PHP_EOL;
            $this->foodCommandHelp();
        }
    }
    
    /**
     * Process the [wash | w] command
     */
    private function washCommand(){
        $name = $this->getParam("name=", "([a-zA-Z]+)");

        if ($name != null) {
            $this->farm->wash($name);
        } else {
            echo "Error!: A name is required to give a shower to a animal". PHP_EOL;
            $this->washCommandHelp();
        }
    }
    
    /**
     * Process the [alive | a] command
     */
    private function aliveCommand() {
        $name = $this->getParam("name=", "([a-zA-Z]+)");

        if ($name != null) {
            $this->farm->alive($name);
        } else {
            echo "Error!: A name is required to show if a animal is alive". PHP_EOL;
            $this->aliveCommandHelp();
        }
    }
    
    /**
     * Returns the queried param using regex
     *
     * @param value variable name to search
     * @param dataType variable pattern equivalent
     * @return param, the param value if a match occurs
     */
    private function getParam(String $value, String $dataType) {
        $param = null;

        if ($this->dataLength > 2) {
            $dataString = join(";", array_slice($this->data, 2));
            $matches = array();
            preg_match('/' . $value . $dataType . '/', $dataString, $matches);
            if (count($matches) > 0) {
                $param = str_replace($value, '', $matches[0]);
            }
        }

        return $param;
    }

    /**
     * Shows help for the [create | c] command
     */
    private function createCommandHelp() {
        echo "The PHP Farm - a simple command line animals app" . PHP_EOL;
        echo PHP_EOL;
        echo "Command Usage:" . PHP_EOL;
        echo "  [create  | c]   <name=> <type=>    creates a animal with name" . PHP_EOL;
        echo PHP_EOL;
        echo "Examples:" . PHP_EOL;
        echo "  php Main.php [create  | c]  name=my_animal_name" . PHP_EOL;
        echo "  php Main.php [create  | c]  name=my_animal_name type=pig" . PHP_EOL;
    }

    /**
     * Shows help for the [delete | c] command
     */
    private function deleteCommandHelp(){
        echo "The PHP Farm - a simple command line animals app" . PHP_EOL;
        echo PHP_EOL;
        echo "Command Usage:" . PHP_EOL;
        echo "  [delete  | d]   <name=>        deletes a animal by name" . PHP_EOL;
        echo "" . PHP_EOL;
        echo "Examples:" . PHP_EOL;
        echo "  php Main.php [delete  | d]  name=my_animal_name" . PHP_EOL;
    }
    
    /**
     * Shows help for the [search | s] command
     */
    private function searchCommandHelp(){
        echo "The PHP Farm - a simple command line animals app" . PHP_EOL;
        echo PHP_EOL;
        echo "Command Usage:" . PHP_EOL;
        echo "  [search  | s]   <name=>       searchs a animal by name" . PHP_EOL;
        echo PHP_EOL;
        echo "Examples:" . PHP_EOL;
        echo "  php Main.php [search  | s]  name=my_animal_name" . PHP_EOL;
    }
    
    /**
     * Shows help for the [food | f] command
     */
    private function foodCommandHelp(){
        echo "The PHP Farm - a simple command line animals app"  . PHP_EOL;
        echo PHP_EOL;
        echo "Command Usage:"  . PHP_EOL;
        echo "  [food  | f]   <name=>       gives food to a animal by name"  . PHP_EOL;
        echo PHP_EOL;
        echo "Examples:"  . PHP_EOL;
        echo "  php Main.php [food  | f]  name=my_animal_name"  . PHP_EOL;
    }
    
    /**
     * Shows help for the [wash | w] command
     */
    private function washCommandHelp(){
        echo "The PHP Farm - a simple command line animals app"  . PHP_EOL;
        echo PHP_EOL;
        echo "Command Usage:"  . PHP_EOL;
        echo "  [wash  | w]   <name=>       gives a shower to a animal by name"  . PHP_EOL;
        echo PHP_EOL;
        echo "Examples:"  . PHP_EOL;
        echo "  php Main.php [wash  | w]  name=my_animal_name"  . PHP_EOL;
    }
    
    /**
     * Shows help for the [alive | a] command
     */
    private function aliveCommandHelp(){
        echo "The PHP Farm - a simple command line animals app"  . PHP_EOL;
        echo PHP_EOL;
        echo "Command Usage:"  . PHP_EOL;
        echo "  [alive  | a]   <name=>       checks if an animal is alive by name"  . PHP_EOL;
        echo PHP_EOL;
        echo "Examples:"  . PHP_EOL;
        echo "  php Main.php [alive  | a]  name=my_animal_name"  . PHP_EOL;
    }
    
    /**
     * Shows help for the app
     */
    private function help() {
        echo "The PHP Farm - a simple command line animals app" . PHP_EOL;
        echo PHP_EOL;
        echo "Usage:" . PHP_EOL;
        echo "  php Main.php [command]" . PHP_EOL;
        echo PHP_EOL;
        echo "Available Commands:" . PHP_EOL;
        echo "  [list    | l]                      list all available animals" . PHP_EOL;
        echo "  [create  | c]   <name=> <type=>    create a animal with name" . PHP_EOL;
        echo "  [delete  | d]   <name=>            delete a animal" . PHP_EOL;
        echo "  [search  | s]   <name=>            search a animal" . PHP_EOL;
        echo "  [food    | f]   <name=>            give food to a animal" . PHP_EOL;
        echo "  [wash    | w]   <name=>            give a shower to a animal" . PHP_EOL;
        echo "  [alive   | a]   <name=>            show if a animal is alive" . PHP_EOL;
        echo "  [help    | h]                      help about commands" . PHP_EOL;
        echo PHP_EOL;
        echo "Flags:" . PHP_EOL;
        echo "  -v, --version  show the app version" . PHP_EOL;
        echo PHP_EOL;
        echo "Use php Main.php [command] --help for more information about a command." . PHP_EOL;
    }
}

?>

File: dotnet/src/Commands.cs
using System;
using System.Text.RegularExpressions;

namespace dotnet
{
    ///<summary>
    /// Class that works as a kind of UI.
    ///</summary>
    public class Commands
    {
        private Farm farm;
        private string[] data;

        ///<summary>
        /// Class constructor, setup the farm
        ///</summary>
        ///<param name="data">The input array.</param>
        public Commands(string[] data)
        {
            this.farm = new Farm();
            this.data = data;
        }

        ///<summary>
        /// Obtain the command and process the input data
        ///</summary>
        public void Process()
        {
            String command = this.GetCommand();
            if (command != null)
            {
                this.VerifyCommand(command);
            }
            else
            {
                this.Help();
            }
        }

        ///<summary>
        /// Returns the command
        ///</summary>
        ///<returns>The first item in the input array, if it's given</returns>
        private String GetCommand()
        {
            return data.Length > 0 ? data[0] : null;
        }

        ///<summary>
        /// Identify the specific action to execute and call the respective action
        ///</summary>
        private void VerifyCommand(string command)
        {
            if (Array.Exists(new[] { "list", "l" }, element => element == command))
            {
                this.ListCommand();
            }
            else if (Array.Exists(new[] { "create", "c" }, element => element == command))
            {
                if (this.CreateCommand())
                {
                    this.ListCommand();
                }
            }
            else if (Array.Exists(new[] { "delete", "d" }, element => element == command))
            {
                if (this.DeleteCommand())
                {
                    this.ListCommand();
                }
            }
            else if (Array.Exists(new[] { "search", "s" }, element => element == command))
            {
                this.SearchCommand();
            }
            else if (Array.Exists(new[] { "food", "f" }, element => element == command))
            {
                this.FoodCommand();
            }
            else if (Array.Exists(new[] { "wash", "w" }, element => element == command))
            {
                this.WashCommand();
            }
            else if (Array.Exists(new[] { "alive", "a" }, element => element == command))
            {
                this.AliveCommand();
            }
            else if (Array.Exists(new[] { "help", "h" }, element => element == command))
            {
                this.Help();
            }
            else
            {
                this.Help();
            }
        }

        ///<summary>
        ///  Process the [list | l] command
        ///</summary>
        private void ListCommand()
        {
            string age = this.GetParam("age=", "(\\d+)");
            string type = this.GetParam("type=", "([a-zA-Z]+)");

            this.farm.ListAnimals(age, type);
        }

        ///<summary>
        ///  Process the [create | c] command
        ///</summary>
        ///<returns>true, if the animal was created</returns>
        private bool CreateCommand()
        {
            string name = this.GetParam("name=", "([a-zA-Z]+)");
            string type = this.GetParam("type=", "([a-zA-Z]+)");

            bool created = false;

            if (!string.IsNullOrEmpty(name))
            {
                Console.WriteLine("Name: create Animal");
                created = this.farm.CreateAnimal(name, type);
            }
            else
            {
                Console.WriteLine("Error!: A name is required to create an animal");
                this.CreateCommandHelp();
            }

            return created;
        }

        ///<summary>
        ///  Process the [delete | c] command
        ///</summary>
        ///<returns>true if the animal was deleted</returns>
        private bool DeleteCommand()
        {
            string name = this.GetParam("name=", "([a-zA-Z]+)");

            bool deleted = false;

            if (!string.IsNullOrEmpty(name))
            {
                deleted = this.farm.DeleteAnimal(name);
            }
            else
            {
                Console.WriteLine("Error!: A name is required to delete a animal");
                this.DeleteCommandHelp();
            }

            return deleted;
        }

        ///<summary>
        ///  Process the [search | s] command
        ///</summary>
        private void SearchCommand()
        {
            string name = this.GetParam("name=", "([a-zA-Z]+)");

            if (!string.IsNullOrEmpty(name))
            {
                this.farm.Search(name);
            }
            else
            {
                Console.WriteLine("Error!: A name is required to search a animal");
                this.SearchCommandHelp();
            }
        }

        ///<summary>
        ///  Process the [food | f] command
        ///</summary>
        private void FoodCommand()
        {
            string name = this.GetParam("name=", "([a-zA-Z]+)");

            if (name != null)
            {
                this.farm.Food(name);
            }
            else
            {
                Console.WriteLine("Error!: A name is required to give food to a animal");
                this.FoodCommandHelp();
            }
        }

        ///<summary>
        ///  Process the [wash | w] command
        ///</summary>
        private void WashCommand()
        {
            string name = this.GetParam("name=", "([a-zA-Z]+)");

            if (!string.IsNullOrEmpty(name))
            {
                this.farm.Wash(name);
            }
            else
            {
                Console.WriteLine("Error!: A name is required to give a shower to a animal");
                this.WashCommandHelp();
            }
        }

        ///<summary>
        ///  Process the [alive | a] command
        ///</summary>
        private void AliveCommand()
        {
            string name = this.GetParam("name=", "([a-zA-Z]+)");

            if (!string.IsNullOrEmpty(name))
            {
                this.farm.Alive(name);
            }
            else
            {
                Console.WriteLine("Error!: A name is required to show if a animal is alive");
                this.AliveCommandHelp();
            }
        }

        ///<summary>
        /// Returns the queried param using regex
        ///</summary>
        ///<param name="values">variable name to search.</param>
        ///<param name="dataType">variable pattern equivalent.</param>
        ///<returns>param, the param value if a match occurs</returns>
        private string GetParam(string value, string dataType)
        {
            string param = null;

            if (this.data.Length > 1)
            {
                string dataString = String.Join(";", this.data);
                string pattern = value+dataType;
                Regex rgx = new Regex(pattern);
                MatchCollection matches = rgx.Matches(dataString);

                if(matches.Count > 0){
                    var result = matches[0].Value.Replace(value, "");
                    if(!String.IsNullOrEmpty(result)){
                        param = result;
                    }
                }
            }

            return param;
        }

        ///<summary>
        /// Shows help for the [create | c] command
        ///</summary>
        private void CreateCommandHelp()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine();
            Console.WriteLine("Command Usage:");
            Console.WriteLine("  [create  | c]   <name=> <type=>    creates a animal with name");
            Console.WriteLine("");
            Console.WriteLine("Examples:");
            Console.WriteLine("  dotnet run [create  | c] name=my_animal_name ");
            Console.WriteLine("  dotnet run [create  | c] name=my_animal_name type=pig ");
        }

        ///<summary>
        /// Shows help for the [delete | c] command
        ///</summary>
        private void DeleteCommandHelp()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine();
            Console.WriteLine("Command Usage:");
            Console.WriteLine("  [delete  | d]   <name=>        deletes a animal by name");
            Console.WriteLine("");
            Console.WriteLine("Examples:");
            Console.WriteLine("  dotnet run [delete  | d] name=my_animal_name ");
        }

        ///<summary>
        /// Shows help for the [search | s] command
        ///</summary>
        private void SearchCommandHelp()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine();
            Console.WriteLine("Command Usage:");
            Console.WriteLine("  [search  | s]   <name=>       searchs a animal by name");
            Console.WriteLine("");
            Console.WriteLine("Examples:");
            Console.WriteLine("  dotnet run [search  | s] name=my_animal_name ");
        }

        ///<summary>
        /// Shows help for the [food | f] command
        ///</summary>
        private void FoodCommandHelp()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine();
            Console.WriteLine("Command Usage:");
            Console.WriteLine("  [food  | f]   <name=>       gives food to a animal by name");
            Console.WriteLine("");
            Console.WriteLine("Examples:");
            Console.WriteLine("  dotnet run [food  | f] name=my_animal_name ");
        }

        ///<summary>
        /// Shows help for the [wash | w] command
        ///</summary>
        private void WashCommandHelp()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine();
            Console.WriteLine("Command Usage:");
            Console.WriteLine("  [wash  | w]   <name=>       gives a shower to a animal by name");
            Console.WriteLine("");
            Console.WriteLine("Examples:");
            Console.WriteLine("  dotnet run [wash  | w] name=my_animal_name ");
        }

        ///<summary>
        /// Shows help for the [alive | a] command
        ///</summary>
        private void AliveCommandHelp()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine();
            Console.WriteLine("Command Usage:");
            Console.WriteLine("  [alive  | a]   <name=>       checks if an animal is alive by name");
            Console.WriteLine("");
            Console.WriteLine("Examples:");
            Console.WriteLine("  dotnet run [alive  | a] name=my_animal_name ");
        }

        ///<summary>
        /// Shows help for the app
        ///</summary>
        private void Help()
        {
            Console.WriteLine("The DotNET Farm - a simple command line animals app");
            Console.WriteLine("");
            Console.WriteLine("Usage:");
            Console.WriteLine("  dotnet run [command] ");
            Console.WriteLine("");
            Console.WriteLine("Available Commands:");
            Console.WriteLine("  [list    | l]   <age=>  <type=>    list all available animals");
            Console.WriteLine("  [create  | c]   <name=> <type=>    create a animal with name");
            Console.WriteLine("  [delete  | d]   <name=>            delete a animal");
            Console.WriteLine("  [search  | s]   <name=>            search a animal");
            Console.WriteLine("  [food    | f]   <name=>            give food to a animal");
            Console.WriteLine("  [wash    | w]   <name=>            give a shower to a animal");
            Console.WriteLine("  [alive   | a]   <name=>            show if a animal is alive");
            Console.WriteLine("  [help    | h]                      help about commands");
            Console.WriteLine("");
            Console.WriteLine("Flags:");
            Console.WriteLine("Use dotnet run help  for more information about a command.");
        }

    }
}
